### ISSUE TITLE
wrong solution with scipy.optimize.linprog

### ISSUE THREAD

**Issue by martinResearch** [scipy/scipy#6690]
I have a very small linear program that causes  scipy.optimize.linprog to give a unexpected result that 
violates the bound constraints despite the status is 0 (Optimization terminated successfully)

the code to reproduce the problem:
    import numpy as np
    import scipy.optimize

```
A_eq=np.array([[ 0.  ,  0.  ,  0.  ,  0.93,  0.  ,  0.65,  0.  ,  0.  ,  0.83,  0.  ]])   
b_eq=np.array([ 0.9626])
A_ub=np.array([[ 0.  ,  0.  ,  0.  ,  1.18,  0.  ,  0.  ,  0.  , -0.2 ,  0.  ,
        -0.22],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
       [ 0.  ,  0.  ,  0.  ,  0.43,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
       [ 0.  , -1.22, -0.25,  0.  ,  0.  ,  0.  , -2.06,  0.  ,  0.  ,
         1.37],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  , -0.25,  0.  ,  0.  ]]) 
b_ub= np.array([ 0.615,  0.   ,  0.172, -0.869, -0.022])
bounds=  np.array([[-0.84, -0.97,  0.34,  0.4 , -0.33, -0.74,  0.47,  0.09, -1.45, -0.73],
               [ 0.37,  0.02,  2.86,  0.86,  1.18,  0.5 ,  1.76,  0.17,  0.32, -0.15]]).T
c=np.array([-1.64,  0.7 ,  1.8 , -1.06, -1.16,  0.26,  2.13,  1.53,  0.66,  0.28])

sol=scipy.optimize.linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, 
                           bounds=bounds, 
                           method='simplex')

x=sol['x']    
print x
#[ 0.37        9.23864382  0.34        0.4         1.18        0.5         0.47
#  0.09        0.32        8.36156603]
print sol['message']
# 'Optimization terminated successfully.'
assert(x[1]<bounds[1][1]+1e-3)
```

the same problem is correlty solved in matlab 

```
A_eq=[[ 0.  ,  0.  ,  0.  ,  0.93,  0.  ,  0.65,  0.  ,  0.  ,  0.83,  0.  ]] 
b_eq=[ 0.9626]
A_ub=[[ 0.  ,  0.  ,  0.  ,  1.18,  0.  ,  0.  ,  0.  , -0.2 ,  0.  , -0.22],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
       [ 0.  ,  0.  ,  0.  ,  0.43,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
       [ 0.  , -1.22, -0.25,  0.  ,  0.  ,  0.  , -2.06,  0.  ,  0.  , 1.37],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  ,  0.  , -0.25,  0.  ,  0.  ]]

b_ub= [ 0.615,  0.   ,  0.172, -0.869, -0.022]
bounds=  [[-0.84, -0.97,  0.34,  0.4 , -0.33, -0.74,  0.47,  0.09, -1.45, -0.73],
           [ 0.37,  0.02,  2.86,  0.86,  1.18,  0.5 ,  1.76,  0.17,  0.32, -0.15]]'
c=[-1.64,  0.7 ,  1.8 , -1.06, -1.16,  0.26,  2.13,  1.53,  0.66,  0.28]

x = linprog(c,A_ub,b_ub,A_eq,b_eq,bounds(:,1),bounds(:,2));


% x =
% 
%     0.3700
%    -0.9700
%     0.3400
%     0.4000
%     1.1800
%     0.5000
%     0.4700
%     0.0900
%     0.3200
%    -0.7300
```

**Comment by martinResearch** [comment]
i'm using scipy version '0.18.0'

**Comment by person142** [comment]
Related to/the same issue as https://github.com/scipy/scipy/issues/5400 I wonder?

**Comment by mdhaber** [comment]
I have another example. 

`import json
from scipy.optimize import linprog
import numpy as np
d = json.loads('{"c": [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], "bounds": [0, 1], "A_eq": [[1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0], [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0], [1.0, 1.0, 0.0, 0.0, 2.0, 2.0, 0.0, 0.0, 3.0, 3.0, 0.0, 0.0, 4.0, 4.0, 0.0, 0.0], [0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 2.0, 2.0, 0.0, 0.0, 3.0, 3.0, 0.0, 0.0, 4.0, 4.0], [1.0, 0.0, 1.0, 0.0, 2.0, 0.0, 2.0, 0.0, 3.0, 0.0, 3.0, 0.0, 4.0, 0.0, 4.0, 0.0], [0.0, 1.0, 0.0, 1.0, 0.0, 2.0, 0.0, 2.0, 0.0, 3.0, 0.0, 3.0, 0.0, 4.0, 0.0, 4.0], [1.0, 0.0, 0.0, 1.0, 2.0, 0.0, 0.0, 2.0, 3.0, 0.0, 0.0, 3.0, 4.0, 0.0, 0.0, 4.0], [0.0, 1.0, 1.0, 0.0, 0.0, 2.0, 2.0, 0.0, 0.0, 3.0, 3.0, 0.0, 0.0, 4.0, 4.0, 0.0]], "b_eq": [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0]}')
sol = linprog(d["c"], A_eq = d["A_eq"], b_eq = d["b_eq"], bounds = d["bounds"])
x = sol["x"]
bounds = d["bounds"]
print sol
print "Bounds satisfied: ", np.logical_and(bounds[0] <= x, x <= bounds[1])`

This is the LP-relaxation of a 0-1 integer program for finding an order-2 magic square. Certainly the integer program is infeasible as no order-2 magic square exists, and I believe the LP-relaxation is infeasible, too.

Note that linprog returns the same solution with or without these bounds and if bounds is [(0,1)]*len(d["c"]). If the bounds are modified (to, say, (-1,1)), the solution changes, but it is still typically incorrect.

**Comment by mdhaber** [comment]
In my second example of this bug, the matrix A_eq is rank-deficient. Removing the redundancy eliminates the problem. As suggested in #7044, potential solutions are to have linprog eliminate the redundancy before solving, warn the user of rank-deficiency, or mention in the documentation that the rows of A_eq need to be linearly independent.

(Note that in the following all row/variable numbers are 0-indexed)

I notice that in the first example of this bug report, row 1 of A_ub and b_ub are empty; this is a trivially redundant constraint. Removing this redundancy alone does _not_ fix the problem, however.

Rows 2 and 4 of the (original) A_ub/b_ub represent simple bounds that are redundant with the bounds argument.

Row 2 represents an upper bound on variable 3 of 0.4. This essentially fixes the value of variable 3 at , because the lower bound is also 0.4. Removing this redundancy and replacing the upper bound with 0.4 did not fix the problem or change the solution.

Row 5 represents a _lower_ bound on variable 7 of 0.088. This is redundant with the lower bound of 0.09 given in the bounds argument. Removing this redundancy did not fix the problem or change the solution.

We are left with one final redundancy, as now both the upper and lower bounds on variable 3 are 0.4, which is really an equality constraint on variable 3. Ideally, variable 3 is eliminated from the problem entirely. 

Once all this is done, linprog returns the same solution as matlab (except that variable 3 has been eliminated, so the objective function value needs 0.4*c[3] added on.

```python
d = json.loads('{"c": [-1.64, 0.7, 1.8, -1.16, 0.26, 2.13, 1.53, 0.66, 0.28], "A_ub": [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -0.2, 0.0, -0.22], [0.0, -1.22, -0.25, 0.0, 0.0, -2.06, 0.0, 0.0, 1.37]], "A_eq": [[0.0, 0.0, 0.0, 0.0, 0.65, 0.0, 0.0, 0.83, 0.0]], "bounds": [[-0.84, 0.37], [-0.97, 0.02], [0.34, 2.86], [-0.33, 1.18], [-0.74, 0.5], [0.47, 1.76], [0.09, 0.17], [-1.45, 0.32], [-0.73, -0.15]], "b_ub": [0.14300000000000002, -0.869], "b_eq": [0.5906]}')
sol=sp.optimize.linprog(d['c'], A_ub=d['A_ub'], b_ub=d['b_ub'], A_eq=d['A_eq'], b_eq=d['b_eq'],bounds=d['bounds'])
print sol.x
```

This prints:
```
[ 0.37 -0.97  0.34  1.18  0.5   0.47  0.09  0.32 -0.73]
```
I can fix this bug, but I need to know which to work on:
Presolve to eliminate redundancy
or
Warn the user of redundancy (and continue or quit)
(the latter is easier, obviously)

At a minimum, the documentation needs to mention that the user must avoid redundancy to ensure an accurate result.

**Comment by rgommers** [comment]
A warning seems warranted either way, preferably followed by eliminating the redundancy and returning the optimal solution. (disclaimer: I'm not an expert. better wait for a second opinion)

**Comment by rgommers** [comment]
@mdhaber can we close this?

**Comment by mdhaber** [comment]
I did not fix this in my PR; I avoided modifying simplex code. So I would guess not because I believe the simplex method still fails this. I can definitely help someone but I cannot do it myself on a schedule right now :-/

**Comment by mdhaber** [comment]
Fixed by #8909
